py/server/tests/test_table.py

```diff
@@ -778,6 +780,66 @@ def test_python_field_access(self):
         self.assertIn("AAPL-GOOG", html_output)
         self.assertIn("2000", html_output)
 
+    def test_vectorization_exceptions(self):
+        t = empty_table(1)
+
+        @dh_vectorize
+        def vectorized_func(p1, p2):
+            return p1 + p2
+
+        def auto_func(p1, p2):
+            return p1 + p2
+
+        @dh_vectorize
+        def no_param_func():
+            return random.randint(0, 100)
+
+        with self.subTest("parameter number mismatch"):
+            with self.assertRaises(DHError) as cm:
+                t1 = t.update("X = vectorized_func(i)")
+            self.assertRegex(str(cm.exception), r".*count.*mismatch", )
+
+            with self.assertRaises(DHError) as cm:
+                t1 = t.update("X = auto_func(i)")
+            self.assertRegex(str(cm.exception), r".*count.*mismatch", )
+
+        with self.subTest("can't cast return value"):
+            with self.assertRaises(DHError) as cm:
+                t1 = t.update("X = (float)vectorized_func(i, ii)")
+            self.assertIn("can't be cast", str(cm.exception))
+
+        with self.subTest("not be part of another expression"):
+            with self.assertRaises(DHError) as cm:
+                t1 = t.update("X = k + vectorized_func(i, ii)")
+            self.assertIn("in another expression", str(cm.exception))
+
+    def test_column_used_twice(self):
+
+        def py_plus(p1, p2) -> int:
+            return p1 + p2
+
+        t = empty_table(1).update("X = py_plus(ii, ii)")
+        self.assertIn("0", t.to_string())
+
+    def test_vectorized_no_arg(self):
+        def py_random() -> int:
+            return random.randint(0, 100)
+
+        t = empty_table(1).update("X = py_random()")
+        self.assertEqual(t.size, 1)
+
+    def test_vectorized_const_arg(self):
+        def py_const(seed) -> int:
+            random.seed(seed)
+            return random.randint(0, 100)
+
+        t = empty_table(1).update("X = py_const(3)")
+        self.assertEqual(t.size, 1)
+
+        seed = 10
+        t = empty_table(1).update("X = py_const(seed)")
+        self.assertEqual(t.size, 1)
```

Comment:
The code that gave me the willies should probably have more tests here to make sure we are getting constants right vs different argument types.